pyscf/lib/numpy_helper.py

```diff
@@ -1177,6 +1206,93 @@ def expm(a):
         y, buf = buf, y
     return y
 
+def omatcopy(a, out=None):
+    """Copies a matrix.
+
+    Parameters
+    ----------
+    a : ndarray
+        Matrix to be copied. The order of the matrix is preserved.
+        a can be either row or column major.
+    out : ndarray, optional
+        Matrix to be overwritten. A new one is allocated if not provided.
+
+    Returns
+    -------
+    out : ndarray
+        Copy of a with the same order.
+    """
+    lda, _, a_cshape = leading_dimension_order(a)
+    if out is None:
+        out = numpy.empty_like(a)
+    ld_out, _, out_cshape = leading_dimension_order(out)
+    assert out_cshape == a_cshape and a_cshape is not None
+    if a.dtype == numpy.double:
+        fn = _np_helper.NPomp_dcopy
+    elif a.dtype == numpy.complex128:
+        fn = _np_helper.NPomp_zcopy
+    else:
+        raise NotImplementedError
+    fn(ctypes.c_size_t(a_cshape[0]),
+       ctypes.c_size_t(a_cshape[1]),
+       a.ctypes.data_as(ctypes.c_void_p),
+       ctypes.c_size_t(lda),
+       out.ctypes.data_as(ctypes.c_void_p),
+       ctypes.c_size_t(ld_out))
+    return out
+
+def zeros(shape, dtype=numpy.double, order='C'):
```

Comment:
np.zeros() should have the pages properly attached to sockets.

Reply:
True. I'll get rid of this then.

Reply:
After some digging, I learned that np.zeros() calls calloc(). It seems the OS does not actually decide allocate the pages until they are accessed for large enough calloc! In other words, if you write to a big array created with np.zeros() using a single thread, it will all stay on the same socket; if you use multiple, it'll be spread out.

First touch can be accomplished with the other routines, so I'm happy to remove this.

Reply:
Thanks for providing the performance benchmark.

The profiling for matmul using np.zeros vs lib.zeros is not directly comparable, since lib.zeros allocate pages out of the `timeit` , while the page allocation cost is included in the tests for np.zeros. Page allocation typically is the most time consuming step when manipulating a large trunk of memory. If you run the profiling second time, the memory allocated by np.zeros or lib.zeros should not have performance difference, as the pages allocated in both cases should be properly attached to numa sockets.

Calloc cannot utilize multiple threads to execute the zeroing operations. lib.zeros is beneficial from the parallel threads for this operation. In this sense, lib.zeros is more favorable given multiple cores become so normal these days. It can be kept here.

Reply:
@sunqm 

Thanks for your feedback!

I ran the profiling again, this time with warm-ups as you suggested. The results are here:
https://gist.github.com/chillenb/e3f533fa2566ca40e48598e32274eb68

The results did not really change much -- they suggest np.zeros is not spreading out pages over sockets on the test system. If you have observed the opposite behavior on your test system, then this is really puzzling! Perhaps the kernel configuration is the cause.